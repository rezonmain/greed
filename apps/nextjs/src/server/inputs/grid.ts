import { GridModel } from "@greed/db";

export const creates = GridModel.pick({
  name: true,
  heightCells: true,
  widthCells: true,
});

export const lists = GridModel.pick({
  name: true,
}).partial({ name: true });

export const updates = GridModel.pick({
  id: true,
  name: true,
  heightCells: true,
  widthCells: true,
}).partial({
  name: true,
  heightCells: true,
  widthCells: true,
});

export const deletes = GridModel.pick({ id: true });
