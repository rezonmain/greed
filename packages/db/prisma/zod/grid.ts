import * as z from "zod"
import { CompleteCell, RelatedCellModel, CompleteGridMeta, RelatedGridMetaModel, CompletePermission, RelatedPermissionModel, CompleteUser, RelatedUserModel } from "./index"

export const GridModel = z.object({
  id: z.string(),
  createdAt: z.date(),
  updatedAt: z.date(),
  gridMetaId: z.string(),
  userId: z.string(),
})

export interface CompleteGrid extends z.infer<typeof GridModel> {
  cells: CompleteCell[]
  GridMeta: CompleteGridMeta
  permissions: CompletePermission[]
  user: CompleteUser
}

/**
 * RelatedGridModel contains all relations on your model in addition to the scalars
 *
 * NOTE: Lazy required in case of potential circular dependencies within schema
 */
export const RelatedGridModel: z.ZodSchema<CompleteGrid> = z.lazy(() => GridModel.extend({
  cells: RelatedCellModel.array(),
  GridMeta: RelatedGridMetaModel,
  permissions: RelatedPermissionModel.array(),
  /**
   * The user who owns the grid
   */
  user: RelatedUserModel,
}))
