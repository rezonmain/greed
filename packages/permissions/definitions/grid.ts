import { type DefinitionOf } from "../types/DefinitionOf";
import { type PermissionsOf } from "../types/PermissionsOf";

export const operations = ["create", "read", "update", "delete"] as const;
export const protectedResources = ["grid", "permission", "approval"] as const;

export type GridPermissionType = PermissionsOf<
  typeof protectedResources,
  typeof operations
>;

export type GridRoles = "grid.owner" | "grid.moderator" | "grid.buyer";

export const gridPermissions = Object.fromEntries(
  protectedResources
    .map((pr) => operations.map((op) => `${pr}.${op}`))
    .flat(1)
    .map((name, i) => [name, 1 << i])
) as DefinitionOf<typeof protectedResources, typeof operations>;

export const gridRolesPermissions: Readonly<
  Record<GridRoles, GridPermissionType[]>
> = {
  "grid.owner": Object.keys(gridPermissions) as GridPermissionType[],
  "grid.moderator": [
    "grid.read",
    "grid.update",
    "permission.read",
    "approval.read",
    "approval.update",
    "approval.create",
    "approval.delete",
  ],
  "grid.buyer": ["grid.read", "approval.read", "approval.create"],
};
