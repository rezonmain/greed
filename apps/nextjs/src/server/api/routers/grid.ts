import { Prisma } from "@greed/db";
import { TRPCError } from "@trpc/server";
import { createTRPCRouter, protectedProcedure } from "~/server/api/trpc";
import { creates, deletes, lists, updates } from "~/server/inputs/grid";

export const gridRouter = createTRPCRouter({
  gridList: protectedProcedure.query(async ({ ctx }) => {
    return await ctx.prisma.grid.findMany({
      where: {
        userId: ctx.session.user.id,
      },
    });
  }),

  gridByName: protectedProcedure.input(lists).query(async ({ ctx, input }) => {
    return await ctx.prisma.grid.findMany({
      where: {
        userId: ctx.session.user.id,
        name: {
          contains: input.name,
        },
      },
    });
  }),

  gridCreate: protectedProcedure
    .input(creates)
    .mutation(async ({ ctx, input }) => {
      try {
        await ctx.prisma.grid.create({
          data: {
            ...input,
            userId: ctx.session.user.id,
          },
        });
      } catch (error) {
        if ((<Prisma.PrismaClientKnownRequestError>error).code === "P2002") {
          throw new TRPCError({
            code: "BAD_REQUEST",
            message: "Please choose a different name for your grid.",
            cause: error,
          });
        }
        throw error;
      }
    }),

  gridUpdate: protectedProcedure
    .input(updates)
    .mutation(async ({ ctx, input }) => {
      try {
        await ctx.prisma.grid.update({
          where: {
            id: input.id,
          },
          data: {
            ...input,
            userId: ctx.session.user.id,
          },
        });
      } catch (error) {
        if ((<Prisma.PrismaClientKnownRequestError>error).code === "P2025") {
          throw new TRPCError({
            code: "NOT_FOUND",
            message: "The grid you're trying to update does not exist.",
            cause: error,
          });
        }
        throw error;
      }
    }),

  gridDelete: protectedProcedure
    .input(deletes)
    .mutation(async ({ ctx, input }) => {
      try {
        await ctx.prisma.grid.delete({
          where: {
            id: input.id,
          },
        });
      } catch (error) {
        if ((<Prisma.PrismaClientKnownRequestError>error).code === "P2025") {
          throw new TRPCError({
            code: "NOT_FOUND",
            message: "The grid you're trying to delete does not exist.",
            cause: error,
          });
        }
        throw error;
      }
    }),
});
