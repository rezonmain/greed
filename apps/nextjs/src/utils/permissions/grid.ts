import { GridPermission } from "@greed/permissions";
import { type GridPermissionType } from "@greed/permissions/definitions/grid";
import { TRPCError } from "@trpc/server";
import { type TRPCProtectedContext } from "~/server/api/trpc";

export const throwIfForbidden = async ({
  ctx,
  gridId,
  message,
  needs,
}: {
  ctx: TRPCProtectedContext;
  gridId: string;
  message: string;
  needs: GridPermissionType;
}) => {
  const serializedPermissions = await getSerializedGridPermissions(ctx, gridId);
  if (!serializedPermissions)
    throw new TRPCError({
      code: "FORBIDDEN",
      message: "Permissions not found.",
    });
  const perm = new GridPermission(serializedPermissions.permissions);
  if (!perm.has(needs))
    throw new TRPCError({
      code: "FORBIDDEN",
      message,
    });
};

const getSerializedGridPermissions = async (
  ctx: TRPCProtectedContext,
  gridId: string
) => {
  return await ctx.prisma.permission.findFirst({
    where: {
      gridId,
      userId: ctx.session.user.id,
    },
    select: {
      permissions: true,
    },
  });
};
