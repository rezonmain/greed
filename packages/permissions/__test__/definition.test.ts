import { describe, expect, it } from "vitest";
import { gridPermissions, gridRolesPermissions } from "../definitions/grid";

describe("permissions/gird permissions definition", () => {
  it("should remain constant", () => {
    expect(gridPermissions).toMatchSnapshot();
  });
});

describe("permissions/roles", () => {
  it("should remain constant", () => {
    expect(gridRolesPermissions).toMatchSnapshot();
  });
});
