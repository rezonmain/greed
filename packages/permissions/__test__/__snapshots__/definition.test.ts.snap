// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`permissions/definition > should remain constant 1`] = `
{
  "approval.create": 256,
  "approval.delete": 2048,
  "approval.read": 512,
  "approval.update": 1024,
  "grid.create": 1,
  "grid.delete": 8,
  "grid.read": 2,
  "grid.update": 4,
  "permission.create": 16,
  "permission.delete": 128,
  "permission.read": 32,
  "permission.update": 64,
}
`;

exports[`permissions/roles > should remain constant 1`] = `
{
  "moderator": [
    "grid.read",
    "grid.update",
    "permission.read",
    "approval.read",
    "approval.update",
    "approval.create",
    "approval.delete",
  ],
  "owner": [
    "grid.create",
    "grid.read",
    "grid.update",
    "grid.delete",
    "permission.create",
    "permission.read",
    "permission.update",
    "permission.delete",
    "approval.create",
    "approval.read",
    "approval.update",
    "approval.delete",
  ],
  "user": [
    "grid.read",
    "approval.read",
    "approval.create",
  ],
}
`;
